### PR TITLE
REF: Method/Visualizer -> Action

### PR DESCRIPTION
**Pull request by jairideout**
Refactored qiime.sdk.Method and qiime.sdk.Visualizer, which previously contained a ton of duplicated code. The classes were also monolithic.

The new design removes Method and Visualizer classes in favor of a general Action class, which wraps a Method/VisualizerCallable. The Callable wraps an appropriate Signature type and is responsible for synchronous/asynchronous execution (including pickling). `from_function`/`from_markdown` now defer to the relevant components for processing these sources. Plugin method/visualizer registration acts as a factory that builds an appropriate Action and registers it with the framework.

### CODE REVIEW

File: qiime/core/callable.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+        raise NotImplementedError
+
+    @classmethod
+    def from_function(cls, function, signature, package):
+        id = function.__name__
+        self = cls.__new__(cls)
+        self._init(package, id, signature, function, ('function', function))
+        return self
+
+    def __init__(self):
```
Comment (jairideout) [inline qiime/core/callable.py]: See if this can be used for unpickling instead of `_init`

File: qiime/core/callable.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        self.id = id
+        self.signature = signature
+        self._callable = callable
+        self._callable_source = callable_source
+
+        if pid is None:
+            pid = os.getpid()
+        self._pid = pid
+
+        self._bind_dynamic_call()
```
Comment (jairideout) [inline qiime/core/callable.py]: Inline these

File: qiime/core/callable.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        callable_wrapper = self._get_callable_wrapper()
+        self._dynamic_call = callable_wrapper
+
+    def _bind_dynamic_async(self):
+        async_wrapper = self._get_async_wrapper()
+        self._dynamic_async = async_wrapper
+
+    def _get_callable_wrapper(self):
+        def callable_wrapper(*args, **kwargs):
+            args = args[1:]
```
Comment (jairideout) [inline qiime/core/callable.py]: Comment that `args[0]` is the function injected by `decorator.decorator`

File: qiime/core/callable.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+            # TODO this may be able to be simplified once Signature respects
+            # order.
+            wrapper_sig = self._callable_sig_converter_(self._callable)
+            wrapper_sig = inspect.Signature.from_callable(wrapper_sig)
+            wrapper_params = wrapper_sig.parameters
+
+            args = {name: value for value, name in zip(args, wrapper_params)}
```
Comment (jairideout) [inline qiime/core/callable.py]: `args` -> `user_input`

File: qiime/core/callable.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+        # Convert the callable's signature into the wrapper's signature and set
+        # it on the wrapper.
+        return decorator.decorator(
+            wrapper, self._callable_sig_converter_(self._callable))
+
+    def _set_wrapper_properties(self, wrapper, name):
+        wrapper.__name__ = wrapper.__qualname__ = name
+        wrapper.__module__ = self.get_import_path()
+        del wrapper.__annotations__
+        del wrapper.__wrapped__
```
Comment (jairideout) [inline qiime/core/callable.py]: Comment this is necessary so that `inspect` doesn't display wrapped function's annotations

File: qiime/core/callable.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+            # caused by MacOSX async runs for now.
+            try:
+                import matplotlib as plt
+                if plt.rcParams['backend'] == 'MacOSX':
+                    raise EnvironmentError(backend_error_template %
+                                           plt.matplotlib_fname())
+            except ImportError:
+                pass
+
+            args = args[1:]
```
Comment (jairideout) [inline qiime/core/callable.py]: Comment why this is necessary

File: qiime/core/callable.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+    def _callable_sig_converter_(self, callable):
+        # No conversion necessary.
+        return callable
+
+    def _callable_executor_(self, callable, view_args, output_types,
+                            provenance):
+        output_views = callable(**view_args)
+        output_views = tuplize(output_views)
+
+        if len(output_views) != len(output_types):
```
Comment (jairideout) [inline qiime/core/callable.py]: Add todo to handle `typing.Tuple` edgecase

File: qiime/core/callable.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+                    "Expected output view type %r, received %r" %
+                    (view_type.__name__, type(output_view).__name__))
+            artifact = qiime.sdk.Artifact._from_view(
+                output_view, semantic_type, provenance)
+            output_artifacts.append(artifact)
+
+        if len(output_artifacts) == 1:
+            return output_artifacts[0]
+        else:
+            return tuple(output_artifacts)
```
Comment (jairideout) [inline qiime/core/callable.py]: make this a `namedtuple`

File: qiime/core/callable.py [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+    # Abstract method implementations:
+
+    def _callable_sig_converter_(self, callable):
+        return DropFirstParameter.from_function(callable)
+
+    def _callable_executor_(self, callable, view_args, output_types,
+                            provenance):
+        # TODO use user-configured temp dir
+        with tempfile.TemporaryDirectory('qiime2-temp-') as temp_dir:
+            ret_val = callable(output_dir=temp_dir, **view_args)
```
Comment (jairideout) [inline qiime/core/callable.py]: pass pathlib obj thru instead of str (when resolving merge conflicts)

File: qiime/__init__.py [hunk 10]
```diff
@@ -12,7 +12,7 @@
 
 __all__ = ['Metadata', 'MetadataCategory', 'Visualization', 'Artifact']
 
-__version__ = '2.0.2'
+__version__ = '2.0.3.dev'
```
Comment (jairideout) [inline qiime/__init__.py]: make this separate PR

File: qiime/core/callable.py (hunk 9 above)
Comment (jairideout) [inline qiime/core/callable.py]: deferring until `OutPath` exists

**Comment by jairideout** [comment]
Group-reviewed by @gregcaporaso, @ebolyen, @thermokarst, and @jakereps.

File: qiime/core/callable.py (hunk 8 above)
Comment (jairideout) [inline qiime/core/callable.py]: done, the parent class (`Callable`) handles this generically

**Comment by jairideout** [comment]
@ebolyen ready for final review/merge!

**Comment by jairideout** [comment]
Added `qiime.plugin.Plugin.actions` property, which returns a read-only dict of all registered `Actions`.

**Comment by jairideout** [comment]
Added `qiime.core.callable.Results` "namedtuple" class and extensive unit tests.

File: qiime/core/callable.py [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+        self = super().__new__(cls, values)
+
+        # Must set `_fields` attribute this way because `__setattr__` prevents
+        # setting directly (necessary for immutability).
+        object.__setattr__(self, '_fields', fields)
+
+        # Idea taken from
+        # http://jfine-python-classes.rtfd.io/en/latest/subclass-tuple.html
+        for i, field in enumerate(fields):
+            setattr(cls, field, property(operator.itemgetter(i)))
```
Comment (jairideout) [inline qiime/core/callable.py]: add a test for this failure (thanks @ebolyen!!!)

File: qiime/core/callable.py [hunk 12]
```diff
@@ -0,0 +1,459 @@
...
+    # `__setattr__` and `__delattr__` must be defined to prevent users from
+    # creating or deleting attributes after this class has been instantiated.
+    # `tuple` and `namedtuple` do not have this problem because they are
+    # immutable (`__slots__ = ()`). We cannot make this class immutable because
+    # we cannot define nonempty `__slots__` when subclassing `tuple`, and we
+    # need the `_fields` attribute. We work around this issue by disallowing
+    # setting and deleting attributes. The error messages here match the error
+    # messages raised by `tuple` in Python 3.5.1 so that this class emulates
+    # `tuple`/`namedtuple`.
+    def __setattr__(self, name, value):
```
Comment (jairideout) [inline qiime/core/callable.py]: use `namedtuple` error message

File: qiime/core/callable.py [hunk 13]
```diff
@@ -0,0 +1,459 @@
...
+        raise AttributeError(
+            "%r object has no attribute %r" % (self.__class__.__name__, name))
+
+    def __delattr__(self, name):
+        raise AttributeError(
```
Comment (jairideout) [inline qiime/core/callable.py]: use `namedtuple` error message

File: qiime/core/callable.py [hunk 14]
```diff
@@ -0,0 +1,459 @@
...
+        return (
+            isinstance(other, Results) and
+            self._fields == other._fields and
+            tuple(self) == tuple(other)
+        )
+
+    def __ne__(self, other):
+        return not (self == other)
+
+    def __repr__(self):
```
Comment (jairideout) [inline qiime/core/callable.py]: use better repr we discussed

**Comment by jairideout** [comment]
@ebolyen @thermokarst this is ready for final review. I've included a temporary commit with the changes I made to address review comments. Let me know if this is GTG and I'll squash commits.

Example of the new `repr` (includes a fixed width field column!):

```
Results (name = value)
------------------------------------------------------------------------------------------
a      = <artifact: SampleData[AlphaDiversity] uuid: 546f3da3-06a0-4f99-a8dd-c9929bf303ec>
bcdefg = <artifact: SampleData[AlphaDiversity] uuid: 546f3da3-06a0-4f99-a8dd-c9929bf303ec>
hi     = <artifact: SampleData[AlphaDiversity] uuid: 546f3da3-06a0-4f99-a8dd-c9929bf303ec>
```

**Comment by ebolyen** [comment]
@jairideout looks good to me!

**Comment by thermokarst** [comment]
Lovin' the new repr. Looks great to me! Thanks for all your hard work on this one.

**Comment by jairideout** [comment]
🎉 🎉 🎉 🎉 🎉 🎉  HUZZAH!!! 🎉 🎉 🎉 🎉 🎉 🎉 

Thanks @ebolyen, @thermokarst, @jakereps, and @gregcaporaso for reviewing this monster PR and for all your helpful comments!

**Comment by gregcaporaso** [comment]
Awesome! Nice work @jairideout!
